### DISCUSSION TITLE
How to change network interface parameters after calling mg_tcpip_init?

### DISCUSSION THREAD

**Discussion by pseudotronics** [cesanta/mongoose#2606] [Ask questions, get answers] [answered]
What's the proper procedure for changing the network interface parameters after calling mg_tcpip_init?

**Comment by scaprile** [comment]
Can you please be more specific ?
What do you want to change ? Can you tell us why ? Can you explain your use case ?

**Reply by pseudotronics** [reply]
Sure.

- Use case 1: Network configuration is set to be DHCP: I want to be able to fallback to a static ip after a period of time.
- Use case 2: User configuration: I want the end user in my products app to be able to change ( dhcp, gateway, static ip, and netmask) according to their installation requirements from within a web interface. So in this instance they would connect via some default configuration to make changes and then reconnect with their new configuration.

The obvious way to do it would be to reset the MCU after configuration changes, but I would rather not have that be solution if it doesn't need to be.

**Reply by scaprile** [reply, marked as answer]
The stack would reconfigure when link goes down and up. So, if you change IP or go static <-> DHCP, you should force that. You should also close all open connections.
The fallback case... Our DHCP client will keep on trying as long as it is configured for DHCP and link is up  (`ifp->state == MG_TCPIP_STATE_UP`). You could probably get away with reconfiguring and setting `ifp->state = MG_TCPIP_STATE_READY`. However, I suggest you reset the interface forcing a link down.
Regarding how to force a link down/up...
https://github.com/cesanta/mongoose/blob/c8d45c9ff1b3042919d19ba823415d35ce1b0fbc/src/net_builtin.c#L849-L861
you should set `ifp->state = MG_TCPIP_STATE_DOWN` so it is seen as up on next PHY poll.

**Reply by pseudotronics** [reply]
Thanks Sergio! 
I will give that a try.
